File: backend/app/database/mongodb/collections/sales_intelligence.py
"""
Beanie MongoDB Document collections for Phase 8: Advanced Sales Intelligence.

Collections:
- SalesIntelligenceReport (Main Beanie Document)
  Contains embedded schemas for:
  - DecisionMaker
  - GrowthSignal
  - CompanyTimeline
  - SalesOpportunityClassification
  - CompanyRelationshipGraph
  - SalesRecommendation
"""
from datetime import datetime, timezone
from typing import Dict, List, Optional
from beanie import Document, Indexed, PydanticObjectId
from pydantic import BaseModel, Field


class DecisionMaker(BaseModel):
    name: str = Field(..., description="Full name of decision maker")
    designation: str = Field(..., description="Job title / role")
    department: str = Field(..., description="Department (Executive, Tech, Sales, Ops, etc.)")
    linkedin_url: Optional[str] = None
    company_email: Optional[str] = None
    personal_email: Optional[str] = None
    phone: Optional[str] = None
    confidence_score: int = Field(80, ge=0, le=100)
    source: str = Field("website_team_page", description="Discovery source signal")
    discovery_timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class GrowthSignal(BaseModel):
    type: str = Field(..., description="hiring | funding | expansion | tech_migration | press | product_launch | social")
    description: str = Field(..., description="Detail of signal detected")
    confidence: int = Field(80, ge=0, le=100)
    source: str = Field("website_scan")
    date: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class Milestone(BaseModel):
    year_or_date: str
    event: str
    category: str = Field("expansion", description="founding | funding | expansion | milestone | event")


class CompanyTimeline(BaseModel):
    founded_year: Optional[str] = None
    expansion_history: List[str] = Field(default_factory=list)
    funding_history: List[str] = Field(default_factory=list)
    milestones: List[Milestone] = Field(default_factory=list)
    current_stage: str = Field("Growth", description="Startup | Growth | Scale-up | Enterprise")
    future_direction: Optional[str] = None
    recent_events: List[str] = Field(default_factory=list)
    ai_summary: Optional[str] = None


class SalesOpportunityClassification(BaseModel):
    categories: List[str] = Field(default_factory=list, description="Hot Opportunity, Enterprise Target, High Growth, etc.")
    primary_category: str = Field("SMB", description="Primary classification category")
    rationale: Optional[str] = None


class SalesGraphNode(BaseModel):
    id: str
    label: str
    type: str = Field(..., description="company | person | campaign | score | intelligence | tech | industry")


class SalesGraphEdge(BaseModel):
    source_id: str
    target_id: str
    relation_type: str = Field(..., description="employs | scored_by | targets | uses_tech | operates_in | exhibits_signal")
    weight: float = 1.0


class CompanyRelationshipGraph(BaseModel):
    nodes: List[SalesGraphNode] = Field(default_factory=list)
    edges: List[SalesGraphEdge] = Field(default_factory=list)


class SalesRecommendation(BaseModel):
    best_contact_person: Optional[str] = None
    best_outreach_channel: str = Field("Email", description="Email | LinkedIn | Phone | Multi-Channel")
    best_time_to_contact: str = Field("Tuesday - Thursday (09:00 - 11:00 AM)")
    pain_points: List[str] = Field(default_factory=list)
    recommended_product_pitch: Optional[str] = None
    conversation_starter: Optional[str] = None
    recommended_email_tone: str = Field("Professional & Consultative")
    risk_factors: List[str] = Field(default_factory=list)
    opportunity_summary: Optional[str] = None
    competitive_advantage: Optional[str] = None
    objections: List[str] = Field(default_factory=list)
    followup_strategy: Optional[str] = None


class SalesIntelligenceReport(Document):
    lead_id: PydanticObjectId
    company_id: Optional[str] = None
    company_name: str
    website_url: Optional[str] = None
    owner_id: PydanticObjectId

    status: str = Field("pending", description="pending | running | completed | failed")
    progress: float = Field(0.0, ge=0.0, le=100.0)
    error_message: Optional[str] = None

    # Intent Scoring (0-100)
    intent_score: int = Field(50, ge=0, le=100)
    intent_level: str = Field("Medium", description="Very Low | Low | Medium | High | Very High")
    intent_reason: Optional[str] = None

    # Embedded Module Data
    decision_makers: List[DecisionMaker] = Field(default_factory=list)
    growth_signals: List[GrowthSignal] = Field(default_factory=list)
    timeline: Optional[CompanyTimeline] = None
    classification: Optional[SalesOpportunityClassification] = None
    graph: Optional[CompanyRelationshipGraph] = None
    recommendations: Optional[SalesRecommendation] = None

    analyzed_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "sales_intelligence_reports"
        indexes = [
            [("owner_id", 1), ("lead_id", 1)],
            [("owner_id", 1), ("status", 1)],
            [("intent_score", -1)],
        ]

File: frontend/src/api/salesIntelligence.ts
import axiosClient from './axiosClient';

export interface DecisionMaker {
  name: string;
  designation: string;
  department: string;
  linkedin_url?: string;
  company_email?: string;
  personal_email?: string;
  phone?: string;
  confidence_score: number;
  source: string;
  discovery_timestamp?: string;
}

export interface GrowthSignal {
  type: string;
  description: string;
  confidence: number;
  source: string;
  date?: string;
}

export interface Milestone {
  year_or_date: string;
  event: string;
  category: string;
}

export interface CompanyTimeline {
  founded_year?: string;
  expansion_history: string[];
  funding_history: string[];
  milestones: Milestone[];
  current_stage: string;
  future_direction?: string;
  recent_events: string[];
  ai_summary?: string;
}

export interface SalesOpportunityClassification {
  categories: string[];
  primary_category: string;
  rationale?: string;
}

export interface SalesGraphNode {
  id: string;
  label: string;
  type: string;
}

export interface SalesGraphEdge {
  source_id: string;
  target_id: string;
  relation_type: string;
  weight: number;
}

export interface RelationshipGraph {
  nodes: SalesGraphNode[];
  edges: SalesGraphEdge[];
}

export interface SalesRecommendation {
  best_contact_person?: string;
  best_outreach_channel: string;
  best_time_to_contact: string;
  pain_points: string[];
  recommended_product_pitch?: string;
  conversation_starter?: string;
  recommended_email_tone: string;
  risk_factors: string[];
  opportunity_summary?: string;
  competitive_advantage?: string;
  objections: string[];
  followup_strategy?: string;
}

export interface SalesIntelligenceStatusResponse {
  id: string;
  lead_id: string;
  company_name: string;
  status: string;
  progress: number;
  intent_score: number;
  intent_level: string;
  error_message?: string;
}

export interface SalesIntelligenceReport {
  id: string;
  lead_id: string;
  company_name: string;
  website_url?: string;
  status: string;
  progress: number;
  intent_score: number;
  intent_level: string;
  intent_reason?: string;
  decision_makers: DecisionMaker[];
  growth_signals: GrowthSignal[];
  timeline?: CompanyTimeline;
  classification?: SalesOpportunityClassification;
  graph?: RelationshipGraph;
  recommendations?: SalesRecommendation;
  analyzed_at?: string;
  created_at: string;
  updated_at: string;
}

export const salesIntelligenceApi = {
  analyzeLead: async (leadId: string): Promise<SalesIntelligenceStatusResponse> => {
    const res = await axiosClient.post<SalesIntelligenceStatusResponse>('/sales-intelligence/analyze', {
      lead_id: leadId,
    });
    return res.data;
  },

  getReportByLead: async (leadId: string): Promise<SalesIntelligenceReport | null> => {
    try {
      const res = await axiosClient.get<SalesIntelligenceReport>(`/sales-intelligence/lead/${leadId}`);
      return res.data;
    } catch (err: any) {
      if (err.response?.status === 404) {
        return null;
      }
      throw err;
    }
  },

  getJobStatus: async (jobId: string): Promise<SalesIntelligenceStatusResponse> => {
    const res = await axiosClient.get<SalesIntelligenceStatusResponse>(`/sales-intelligence/${jobId}/status`);
    return res.data;
  },

  getGrowthSignals: async (leadId: string): Promise<GrowthSignal[]> => {
    const res = await axiosClient.get<GrowthSignal[]>(`/sales-intelligence/${leadId}/signals`);
    return res.data;
  },

  getDecisionMakers: async (leadId: string): Promise<DecisionMaker[]> => {
    const res = await axiosClient.get<DecisionMaker[]>(`/sales-intelligence/${leadId}/decision-makers`);
    return res.data;
  },

  getTimeline: async (leadId: string): Promise<CompanyTimeline> => {
    const res = await axiosClient.get<CompanyTimeline>(`/sales-intelligence/${leadId}/timeline`);
    return res.data;
  },

  getRecommendations: async (leadId: string): Promise<SalesRecommendation> => {
    const res = await axiosClient.get<SalesRecommendation>(`/sales-intelligence/${leadId}/recommendations`);
    return res.data;
  },

  deleteReport: async (leadId: string): Promise<{ status: string; message: string }> => {
    const res = await axiosClient.delete<{ status: string; message: string }>(`/sales-intelligence/lead/${leadId}`);
    return res.data;
  },
};
